### PR TITLE
[NFI]: Simplify declaration of ShapeHelper classes

### PR DESCRIPTION
**Pull request by etiotto**
The declaration of "ShapeHelper" classes for each of the ONNX operators contains repeated boilerplate code. Example:

```
// Shape for ArgMax
struct ONNXArgMaxOpShapeHelper : public ONNXOpShapeHelper<mlir::ONNXArgMaxOp> {
  ONNXArgMaxOpShapeHelper(mlir::ONNXArgMaxOp *newOp);
  ONNXArgMaxOpShapeHelper(mlir::ONNXArgMaxOp *newOp, mlir::OpBuilder *rewriter,
      ArrayValueIndexCapture::GetDenseVal fGetDenseVal,
      ArrayValueIndexCapture::LoadVal fLoadVal);
  mlir::LogicalResult computeShape(mlir::ONNXArgMaxOpAdaptor operandAdaptor);
};

// Shape for Clip.
struct ONNXClipOpShapeHelper : public ONNXOpShapeHelper<mlir::ONNXClipOp> {
  ONNXClipOpShapeHelper(mlir::ONNXClipOp *newOp);
  ONNXClipOpShapeHelper(mlir::ONNXClipOp *newOp, mlir::OpBuilder *rewriter,
      ArrayValueIndexCapture::GetDenseVal fGetDenseVal,
      ArrayValueIndexCapture::LoadVal fLoadVal);
  mlir::LogicalResult computeShape(mlir::ONNXClipOpAdaptor operandAdaptor);
};
```

This PR makes it easier to declare those classes and reduces boilerplate code. Example: 
```
DECLARE_SHAPE_HELPER(ONNXArgMaxOp)
DECLARE_SHAPE_HELPER(ONNXArgMinOp)
```

This PR also implement shape inference for the ~ArgMinOp` ONNX operator. 

Signed-off-by: Ettore Tiotto <etiotto@ca.ibm.com>

### CODE REVIEW

File: src/Dialect/ONNX/ShapeInference/ArgMin.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+LogicalResult ONNXArgMinOpShapeHelper::computeShape(
+    ONNXArgMinOpAdaptor operandAdaptor) {
+  // Get info about input data operand.
+  Value data = operandAdaptor.data();
+  int64_t dataRank = data.getType().cast<ShapedType>().getRank();
+  int64_t axisValue = op->axis();
+
+  // axis attribute must be in the range [-r,r-1], where r = rank(data).
+  if (axisValue < -dataRank || axisValue >= dataRank)
+    return onnx_mlir::Diagnostic::attributeOutOfRange(*op->getOperation(),
```
Comment (chentong319) [inline src/Dialect/ONNX/ShapeInference/ArgMin.cpp:31]: Should the check code be put into verifier?

**Comment by chentong319** [comment]
The previous interface `getResult` intends to handle multiple output for an Op, even though most of the Op has only one output. Make sure the new interface can handle multiple outputs.

File: src/Dialect/ONNX/ShapeInference/ArgMin.cpp (hunk 1 above)
Comment (etiotto) [reply src/Dialect/ONNX/ShapeInference/ArgMin.cpp:31]: Yes, we need to diagnose `axis` out of range both during verification and during shape inference. 
In some cases the rank if the input tensor is not known during verification and so by adding the diagnostic also during shape inference we have a chance to report this problem when we know the shape of the input.

File: src/Dialect/ONNX/ShapeInference/ArgMin.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (whitneywhtsang) [inline src/Dialect/ONNX/ShapeInference/ArgMin.cpp:1]: This file looks identical to ArgMax, anything we can do about that?

File: src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp [hunk 3]
```diff
@@ -63,14 +63,12 @@ template <class OP>
 struct ONNXOpShapeHelper {
   // Constructor for shape inference. Reuse scope if given, otherwise create one
   // now and free it in destructor.
-  ONNXOpShapeHelper(
-      OP *newOp, int numResults, IndexExprScope *inScope = nullptr);
+  ONNXOpShapeHelper(OP *newOp, int numResults, IndexExprScope *inScope);
```
Comment (whitneywhtsang) [inline src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp:66]: What's the reason to remove the default argument?

File: src/Dialect/ONNX/ShapeInference/ArgMin.cpp (hunk 2 above)
Comment (etiotto) [reply src/Dialect/ONNX/ShapeInference/ArgMin.cpp:1]: Good point. I can create a template.

File: src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp (hunk 3 above)
Comment (whitneywhtsang) [reply src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp:66]: I guess you meant 
```
#define DECLARE_SHAPE_HELPER(OpName)                                           \
  class OpName##ShapeHelper : public ONNXOpShapeHelper<mlir::OpName> {         \
  public:                                                                      \
    OpName##ShapeHelper(                                                       \
        mlir::OpName *newOp, IndexExprScope *inScope = nullptr)                \
```

**Review by whitneywhtsang — APPROVED** [review]

**Comment by etiotto** [comment]
@jenkins-droid please test this.

**Comment by etiotto** [comment]
@jenkins-droid test this please

**Pull request merged**